// node-red-contrib-nora/build/nodes/nora-switch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var nora_1 = require("../nora");
var util_1 = require("./util");
module.exports = function (RED) {
    RED.nodes.registerType('nora-switch', function (config) {
        var _this = this;
        RED.nodes.createNode(this, config);
        var noraConfig = RED.nodes.getNode(config.nora);
        if (!noraConfig || !noraConfig.token) {
            return;
        }
        var close$ = new rxjs_1.Subject();
        var on$ = new rxjs_1.BehaviorSubject(false);
        var stateString$ = new rxjs_1.Subject();
        var _a = util_1.convertValueType(RED, config.onvalue, config.onvalueType, { defaultValue: true }), onValue = _a.value, onType = _a.type;
        var _b = util_1.convertValueType(RED, config.offvalue, config.offvalueType, { defaultValue: false }), offValue = _b.value, offType = _b.type;
        var device$ = nora_1.NoraService
            .getService(RED)
            .getConnection(noraConfig, this, stateString$)
            .pipe(operators_1.switchMap(function (connection) { return connection.addDevice(config.id, {
            type: 'switch',
            name: config.devicename,
            roomHint: config.roomhint || undefined,
            state: { online: true, on: on$.value },
        }); }), operators_1.publishReplay(1), operators_1.refCount(), operators_1.takeUntil(close$));
        device$.pipe(operators_1.switchMap(function (d) { return d.errors$; }), operators_1.takeUntil(close$)).subscribe(function (err) { return _this.warn(err); });
        rxjs_1.combineLatest(device$, on$)
            .pipe(operators_1.tap(function (_a) {
            var _ = _a[0], on = _a[1];
            return notifyState(on);
        }), operators_1.skip(1), operators_1.takeUntil(close$))
            .subscribe(function (_a) {
            var device = _a[0], on = _a[1];
            return device.updateState({ on: on });
        });
        device$.pipe(operators_1.switchMap(function (d) { return d.state$; }), operators_1.takeUntil(close$)).subscribe(function (s) {
            var value = s.on;
            notifyState(s.on);
            _this.send({
                payload: util_1.getValue(RED, _this, value ? onValue : offValue, value ? onType : offType),
                topic: config.topic
            });
        });
        this.on('input', function (msg) {
            if (config.passthru) {
                _this.send(msg);
            }
            var myOnValue = util_1.getValue(RED, _this, onValue, onType);
            var myOffValue = util_1.getValue(RED, _this, offValue, offType);
            if (RED.util.compareObjects(myOnValue, msg.payload)) {
                on$.next(true);
            }
            else if (RED.util.compareObjects(myOffValue, msg.payload)) {
                on$.next(false);
            }
        });
        this.on('close', function () {
            close$.next();
            close$.complete();
        });
        function notifyState(on) {
            stateString$.next("(" + (on ? 'on' : 'off') + ")");
        }
    });
};
